4098 Update user name handling

cancelei opened the pull request:
<!--Read comments, before committing pull request read checklist again

# Checklist:

- I have performed a self-review of my own code,
- I have commented my code, particularly in hard-to-understand areas,
- I have made corresponding changes to the documentation,
- I have added tests that prove my fix is effective or that my feature works,
- New and existing unit tests pass locally with my changes ("bundle exec rake"),
- Title include "WIP" if work is in progress.

-->

Resolves #4098  
PS from Author: If you are new to this PR move to the last comments to see the current progress that contains current updates on feedback cycles. More feedback is always appreciated.


### Description
<!-- Please include a summary of the change and which issue is fixed. 
Please also include relevant motivation and context.
Guide questions:
  - What motivated this change (if not already described in an issue)?
  - What alternative solutions did you consider?
  - What are the tradeoffs for your solution?
   
List any dependencies that are required for this change. (gems, js libraries, etc.)

Include anything else we should know about. -->

### Type of change

<!-- Please delete options that are not relevant. -->

* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce. 
Do we need to do anything else to verify your changes? 
If so, provide instructions (including any relevant configuration) so that we can reproduce? -->

I have tested using the user spec used for validating user-related activities. There I added a case when the user name is not filled in, it should be valid and also return the name "Name Not Provided" when accessing the method display_name.

**This PR is ready for a first review.**

Disclaimer:
This PR contains a meaningful commit history where I tried to resolve solely the issue at hand, but when changing the user model breaking test logic appeared. So as a reviewer, I recommend you can follow the steps where I pass all tests with 2 additional file changes (1 in Rails Helper and 1 in an unrelated _spec file). Only with both changes, do all tests pass. Most likely there could be other ways to make all tests pass that could be more suitable.

The test that is not passing in the last commit is passing locally:
<img width="693" alt="image" src="https://github.com/rubyforgood/human-essentials/assets/96393104/7c6c8e51-e92b-452c-9b58-e6e64278f520">

cielf commented on spec/requests/admin_requests_spec.rb:
```diff
@@ -12,19 +12,18 @@
+        expect(response.body).to match(/20 New Users/im)
+        expected_path = edit_admin_user_path(user, organization_id: "admin")
+        expect(response.body).to include(CGI.escapeHTML(expected_path))
       end
```
The changes you made to this test make it not check the email.   I suggest going back to the original test -- except that the user should have a nil name, right?

cancelei replied:
Hello @cielf. I understand that testing the email is our priority when implementing this test. But at the same time, I checked that the Dashboard view doesn't show email on screen, there is an email being used just to bring other information.

I think there are 2 ways:
1) Implementing a change to the dashboard to show user emails. But that probably would require an extra Issue, I guess. I went all the way on this and have tested adding this line to the dashboard and it test passed using the old test (I also added user null as you suggested): 

<div class="users-list-email"><%= user.email %></div>.

/app/view/admin/dashboard.html.erb after line 97.

Then bundle exec rspec ./spec/requests/admin_requests_spec.rb works without changing the original test.

2) Not considering testing the email, since this test was passing (as "False-Positive" test before) because the view doesn't show the email. In that case, I would need to remove the line that tests the email from the original test, since it will not find any email in the view as commented before.

Hope I made myself understandable. I look up to your guidance for the resolution of this case.

cielf replied:
The dashboard currently shows the user email if and only if the user name is "Name Not Provided".  Since we are reworking to get rid of "Name Not Provided" that check should be for "Name Not Provided" or nil.   But we still need the email to be shown in these cases so that we can differentiate between any users who have not provided a name.

cancelei replied:
Thanks @cielf I didn't think of this case you described. So I implemented a new set of changes to the proposed code included in this PR.

cielf commented on spec/requests/admin_requests_spec.rb:
```diff
@@ -9,26 +9,30 @@
-    context "with rendered views" do
-      let!(:users_list) { create_list(:user, 25) }
-      let!(:user) { create(:user, name: "Name Not Provided") }
-      edit_user_path_regex = Regexp.new('admin/users/[0-9]+/edit\\?organization_id=admin')
```
Why did you remove this instead of changing the create user to create a user with a null name?

cancelei replied:
@cielf I thought this part of the test was a bit superficial, so I decided to remove it. But now I have added back in with enhancements and the different scenarios that should be tested including this factor. I also included a test on the user_list itself as well as having meaningful variable names when setting up different test cases.

awwaiid commented on db/schema.rb:
```diff
@@ -250,7 +250,7 @@
     t.float "latitude"
     t.float "longitude"
-    t.index ["latitude", "longitude"], name: "index_donation_sites_on_latitude_and_longitude"
+        t.index ["latitude", "longitude"], name: "index_donation_sites_on_latitude_and_longitude"
```
Weird whitespace

awwaiid commented on spec/models/user_spec.rb:
```diff
@@ -40,10 +40,17 @@
+    it "without a name results in display_name 'Name Not Provided'" do
+      user = build(:user, name: nil)
+      expect(user).to be_valid
+      expect(user.name).to eq(nil)
```
```suggestion
      expect(user.name).to be_nil
```

awwaiid commented on spec/rails_helper.rb:
```diff
@@ -117,6 +117,7 @@ def self.capybara_tmp_path
   config.around(:each, skip_transaction: true) do |example|
     config.use_transactional_fixtures = false
     example.run
+  ensure
```
I suspect this is from an un-caught exception, which we can change into an expectation. I'll help find it.

cancelei replied:
@awwaiid it seems that after removing "ensure", all tests are passing. With that, this PR is ready for review.

awwaiid commented on spec/requests/admin_requests_spec.rb:
```diff
@@ -9,22 +9,51 @@
-      edit_user_path_regex = Regexp.new('admin/users/[0-9]+/edit\\?organization_id=admin')
+      let!(:user_with_name) { create(:user, name: "John Doe", email: "john@example.com") }
+
+      it "displays the users list" do
```
Let's go ahead and drop this block (and `users_list`) -- I don't think we need it compared to the rest of the name/no-name examples.

awwaiid commented:
This looks great overall -- we're going to do a bit of manual testing before merging!

awwaiid commented:
Please merge `main` into this branch, and then apply the changes we talked about for removing the default value in the service object, and then ping @cielf.

cancelei commented:
> Please merge `main` into this branch, and then apply the changes we talked about for removing the default value in the service object, and then ping @cielf.

@cielf First of all have a nice week! I anticipated the conclusion of this PR following the guidance from brook, could you review it and in case there is something to improve, let me know.

cielf commented on spec/services/user_invite_service_spec.rb:
```diff
@@ -94,7 +94,7 @@
     }.to change(ActionMailer::Base.deliveries, :count).by(1)
 
-    expect(result.reload.name).to eq("Name Not Provided")
+    expect(result.display_name).to eq("Name Not Provided")
```
Hmm.   Assuming there was a reason there was a reload in the sequence in the first place,  I would expect to still have the reload in the changed spec.

cielf commented:
Tested the admin dashboard with this -- I'm getting "Name not Provided"  rather than the email there  on a brand new user,  though it does show the email for users that were created with the seed.

cielf commented on spec/services/user_invite_service_spec.rb:
```diff
@@ -79,7 +79,7 @@
     expect(user.name).to eq("Partner User")
-    expect(user).not_to be_nil
+    expect(user).to be
     expect(user).to have_role(Role::PARTNER, @partner)
```
This change doesn't seem necessary.  Though given the order of the checks, I don't think this line is ever going to fail anyway -- if the user doesn't exist, then user.name is certainly not going to equal "Partner User".

cielf commented:
It appears that the User, at least if created from the admin interface,  is getting created with the name "", rather than nil.    That will need to be addressed.

cielf commented:
That also happens if they are created using org_admin1@example.com.   I expect there is an issue in UserInviteService, but it might be in how it's being called.

cielf reviewed (dismissed):
Hey @cancelei This is not behaving in quite the way I expected it to.   From the manual testing,  in the admin dashboard I expect the email to be shown on a newly added user with no name -- that's not happening.    A little digging also showed that the name is getting set to "" rather than nil on a newly added user.

cancelei commented:
> Hey @cancelei This is not behaving in quite the way I expected it to. From the manual testing, in the admin dashboard I expect the email to be shown on a newly added user with no name -- that's not happening. A little digging also showed that the name is getting set to "" rather than nil on a newly added user.

You are right! I added some changes in the views that were blocking adding users without names from the front end, but in the backend, I still haven't found the place where the user name is turning from nil to "". I will be looking into it in the next 24 hours, to complement the current PR, but if you have some hints, I will try to do my best afterward conditioned to my current VS Code-empowered searching capabilities. But I have some ideas on what words to mix to find possible results in the future.

seanmarcia commented on app/models/user.rb:
```diff
@@ -82,6 +81,10 @@ class User < ApplicationRecord
   has_many :submitted_requests, class_name: "Request", foreign_key: :partner_user_id, dependent: :destroy, inverse_of: :partner_user
 
+  def display_name
+    name.presence || "Name Not Provided"
```
TIL about .presence. Thank you :)

cancelei commented:
Hello @cielf, I did some improvements in this PR:

The biggest change was creating a _spec file for dashboard_helper that contains the recently_added_user_display_text which was something you mentioned and we didn't have before. So we can track the different scenarios for this when there is a name when the name is nil, and when the name is blank as ("").

I refactored the tests to gain robustness and added a new one that uses the "name is blank as ("")" since the submitted forms with blank field originate "" and not nil as I was expecting previously. That's why the previous tests were passing, I was expecting the name to be nil and not "".

The current CI run fails due to a FERRUM error (possibly a flaky test).

cielf commented:
I have this on my list to manually test tomorrow.  There's enough new code (including some things I think are redundant) that I'm going to ask @awwaiid  to take another look.

cielf commented:
LGTM from a what the users see and do  p.o.v.

cancelei commented:
> LGTM from a what the users see and do p.o.v.

@awwaiid let me know if there is anything I could remove or add to fulfill the definition of done in this PR.
